### ISSUE TITLE
`Could not connect to the Radicle daemon` on `rad project checkout ...` with both daemons running

### ISSUE THREAD

**Issue by hxrts** [radicle-dev/radicle#587]
reported by [@Bag_End_](https://twitter.com/Bag_End_) on twitter

## system details

- Ubuntu installed via apt
- not behind firewall
- IPFS connects when run for non-radicle purposes

## terminal output

`rad daemon-ipfs` output:

```
Initializing daemon... go-ipfs version: 0.4.18- Repo version: 7 System version: amd64/linux Golang version: go1.11.1 Successfully raised file descriptor limit to 2048. Swarm is limited to private network of peers with the swarm key Swarm key fingerprint: 9bbc6ea04403a4595876eedfadb1aa93 Swarm listening on /ip4/127.0.0.1/tcp/9303 Swarm listening on /ip4/172.17.0.1/tcp/9303 Swarm listening on /ip4/172.18.0.1/tcp/9303 Swarm listening on /ip4/172.19.0.1/tcp/9303 Swarm listening on /ip4/172.20.0.1/tcp/9303 Swarm listening on /ip4/192.168.1.18/tcp/9303 Swarm listening on /ip6/::1/tcp/9303 Swarm listening on /p2p-circuit Swarm announcing /ip4/127.0.0.1/tcp/9303 Swarm announcing /ip4/172.17.0.1/tcp/9303 Swarm announcing /ip4/172.18.0.1/tcp/9303 Swarm announcing /ip4/172.19.0.1/tcp/9303 Swarm announcing /ip4/172.20.0.1/tcp/9303 Swarm announcing /ip4/192.168.1.18/tcp/9303 Swarm announcing /ip6/::1/tcp/9303 API server listening on /ip4/127.0.0.1/tcp/9301 Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/9302 Daemon is ready
```

`rad daemon-radicle` output:

```
INFO    Start listening port=8909
INFO    Following as reader machine-id=12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn current-input-index=MachineEntryIndex zdpuB1dS7MzsbQnZ7hM1CqoQKG6eXcjmjHUh3VqFKnP3CRf4c
```

`rad project checkout 12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn` output:

```
Could not connect to the Radicle daemon, most likely it is not running. Please see the documentation for how to start it.
```

**Comment by jameshaydon** [comment]
Another user was having the same problem, but was able to confirm with `curl` that the daemon was responding to requests.

**Comment by hxrts** [comment]
hmm, interesting. tagging @WaughB, who originally reported, in case they want to keep tabs.

**Comment by rschulman** [comment]
I have a perhaps related problem, but in my case the radicle daemon won't start at all, saying it can't reach the IPFS daemon. IPFS is running, however, and responds to `curl`. Happy to help with any further debugging.

**Comment by xla** [comment]
@rschulman Are you running the radicle ipfs daemon or the general one?

**Comment by rschulman** [comment]
I believe I am running the radicle one. I'll check again to be sure when I get back home.

**Comment by rschulman** [comment]
Having just checked, yes, I am running the radicle-ipfs daemon and only that one.

**Comment by christroutner** [comment]
I'm experiencing the same issue. Trying to install and use radicle for the first time.

**Comment by frankier** [comment]
Same on Debian

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux buster/sid
Release:	testing
Codename:	buster

**Comment by Lupino** [comment]
Same on macos (I compile with whe master branch)

**Comment by jameshaydon** [comment]
This PR: https://github.com/radicle-dev/radicle/pull/603
will print out more debugging information when the connection cannot be established.

**Comment by jkarni** [comment]
What version of `ipfs` is this on? (You can get it with `rad ipfs version`)

**Comment by hxrts** [comment]
original issue was reported while running go-ipfs `0.4.18`

**Comment by christroutner** [comment]
I was using the latest version of IPFS: ipfs version 0.4.19

**Comment by Lupino** [comment]
ipfs version 0.4.19

**Comment by iinstances** [comment]
Same problem here reproduced in 3 differents hosts.

### Main terminal
`➜  **rad project checkout 12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn**

_Could not connect to the Radicle daemon, most likely it is not running. Please see the documentation for how to start it._

➜  radicle-playground mkdir acme && cd acme

➜  **rad project init**
? What's the name of your project: acme
? Briefly describe the project: This is my description
_Could not connect to the Radicle daemon, most likely it is not running. Please see the documentation for how to start it._

➜  **rad ipfs version**
ipfs version 0.4.19
➜  **lsb_release -a**
No LSB modules are available.
Distributor ID: LinuxMint
Description:    Linux Mint 19 Tara
Release:        19
Codename:       tara`

### ipfs daemon terminal
**ipfs daemon**

Initializing daemon...
go-ipfs version: 0.4.19-
Repo version: 7
System version: amd64/linux
Golang version: go1.11.5
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/172.24.236.29/tcp/4001
Swarm listening on /ip4/192.168.1.41/tcp/4001
Swarm listening on /ip6/::1/tcp/4001
Swarm listening on /ip6/fc25:7fba:d97a:14fd:ad25::1/tcp/4001
Swarm listening on /p2p-circuit
Swarm announcing /ip4/127.0.0.1/tcp/4001
Swarm announcing /ip4/172.24.236.29/tcp/4001
Swarm announcing /ip4/192.168.1.41/tcp/4001
Swarm announcing /ip6/::1/tcp/4001
Swarm announcing /ip6/fc25:7fba:d97a:14fd:ad25::1/tcp/4001
API server listening on /ip4/127.0.0.1/tcp/5001
WebUI: http://127.0.0.1:5001/webui
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
Daemon is ready


### rad daemon-ipfs terminal
**rad daemon-ipfs**                          

Initializing daemon...
go-ipfs version: 0.4.19-
Repo version: 7
System version: amd64/linux
Golang version: go1.11.5
Swarm is limited to private network of peers with the swarm key
Swarm key fingerprint: 9bbc6ea04403a4595876eedfadb1aa93
Swarm listening on /ip4/127.0.0.1/tcp/9303
Swarm listening on /ip4/172.24.236.29/tcp/9303
Swarm listening on /ip4/192.168.1.41/tcp/9303
Swarm listening on /ip6/::1/tcp/9303
Swarm listening on /ip6/fc25:7fba:d97a:14fd:ad25::1/tcp/9303
Swarm listening on /p2p-circuit
Swarm announcing /ip4/127.0.0.1/tcp/9303
Swarm announcing /ip4/172.24.236.29/tcp/9303
Swarm announcing /ip4/192.168.1.41/tcp/9303
Swarm announcing /ip6/::1/tcp/9303
Swarm announcing /ip6/fc25:7fba:d97a:14fd:ad25::1/tcp/9303
API server listening on /ip4/127.0.0.1/tcp/9301
WebUI: http://127.0.0.1:9301/webui
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/9302
Daemon is ready


### rad daemon-radicle terminal
**rad daemon-radicle**
INFO    Start listening port=8909
INFO    Following as reader machine-id=12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn current-input-index=MachineEntryIndex zdpuB1dS7MzsbQnZ7hM1CqoQKG6eXcjmjHUh3VqFKnP3CRf4c
ERROR   Could not create IPFS machine message=Timeout communicating with IPFS daemon api-path=name/publish

**Still NOT sure if only one ipfs daemon is needed**

**Comment by jkarni** [comment]
> INFO Following as reader machine-id=12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn current-input-index=MachineEntryIndex zdpuB1dS7MzsbQnZ7hM1CqoQKG6eXcjmjHUh3VqFKnP3CRf4c
ERROR Could not create IPFS machine message=Timeout communicating with IPFS daemon api-path=name/publish

This is interesting. If `daemon-radicle` tried to both read and create a machine, that means it did receive messages and so could be connected to. It seems like the error message we're sending out is not appropriate.

Timeout on `name/publish` seems to be a consequence of the fact that our network is growing a lot, and the whole IPNS thing fares poorly on larger networks, so the timeouts we had before the release might be too short now. (I imagine when reading the timeout problem is also with IPNS, but on resolution.) I guess one makeshift solution is to increase all of them (@jameshaydon already did so for pubsub). In the longer term that's not great in terms of user experience, of course.

**Comment by jkarni** [comment]
@iinstances just to help validate my thoughts, can you try starting both daemons, waiting till they're both ready (and maybe another 30 seconds to make sure the daemon had time to poll), and then running the command `rad project checkout 12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn` again? I would assume that it should work then.

**Comment by iinstances** [comment]
@jkarni, I have tried again just as you say, both daemons, wait more thann 30 seconds, but, also;

`rad project checkout 12D3KooWRsrUcqCFU3ompktgTo197gguibN3cjmgaoQy8AS54iHn
Could not connect to the Radicle daemon, most likely it is not running. Please see the documentation for how to start it.`
:(

**Comment by jameshaydon** [comment]
Good news: we did lots of debugging and think https://github.com/radicle-dev/radicle/pull/617 is a fix.

**Comment by christroutner** [comment]
If someone wants to produce a .deb file with PR #617, I'd love to try it out and test it.

**Comment by Lupino** [comment]
the same error on macos when i test with #617 

```
INFO    Acting as writer machine-id=12D3KooWPWvcEJyVu44sucrVLDXv2S2ziL2oGe9fMLPNYLpv3Xi8
INFO    Start listening port=8909
INFO    Following as reader machine-id=12D3KooWPS3UXcvSZSXfi7P4J9Ut8MMVNvN63HHiCSP8rxj3RmtC current-input-index=MachineEntryIndex zdpuApyuhXuWYQEmBai97Kw7ZRWRJ8evJdm9zfoCY1YQLPtuw
INFO    Created new IPFS machine machine-id=12D3KooWMpRshXKr3oru2Cr73AA1gzyGnZQ969gbjJTfd2eWbaBa
INFO    Acting as writer machine-id=12D3KooWMpRshXKr3oru2Cr73AA1gzyGnZQ969gbjJTfd2eWbaBa
```

**Comment by alanz** [comment]
I just built #617 locally, and can confirm that it works for me. On Debian Testing.

**Comment by alanz** [comment]
Except when I try to propose a patch it fails

```
$ rad patch propose 526dc05156482c430dbd16bf886abdcbe89aecff
No response received from the machine owner. This could be because they are offline, or that waiting for the response timed out:
  machine-id: 12D3KooWAmB1gB9kceaGfKaUHfhcmfUzXdJEbzRnT73mF1DrknSk
```

**Comment by iinstances** [comment]
Update with better news!

* Remove .deb package: `apt remove radicle`
* Git source updated - `git pull`
* Built master branch: `stack build`
* Local installation with: `stack install :rad :radicle`
* Create a test directory acme: `mkdir acme; cd acme`

With system wide ipfs daemon stopped run separately the two radicle daemons;
* `rad daemon-ipfs`
* `rad daemon-radicle`

And then:

**rad project init**
? What's the name of your project: radicle-mine
? Briefly describe the project: A Test
 Initialized Git repository empty in ../acme/.git/
? What kind of repository would you like to use?
1. New peer-to-peer repository (git-ipfs **EXPERIMENTAL**)
2. Add your own remote (e.g.: github / gitlab / ...)
1
=> Assembled rad-issue machine
=> Assembled rad-patch machine
=> project id: 12D3KooWCsVoE9FcoyweTZpquDvP7TH4uzR7wq88ec55J32Ci865
=> adding "origin" remote: ipfs://ipns/12D3KooWNwPGi9yQs5PyqsL569VUQtQghFNgNug11qZ4WzTAx5VD
[master (commit-raíz) 8236223] Radicle Setup
Resolving IPNS name 12D3KooWNwPGi9yQs5PyqsL569VUQtQghFNgNug11qZ4WzTAx5VD
Pushed 1 objects
Updating IPNS link 12D3KooWNwPGi9yQs5PyqsL569VUQtQghFNgNug11qZ4WzTAx5VD to /ipfs/QmPsYV8QJn11JP2tPizScDVAZxKSzUoBN9aLPwqRDyFSNo
To ipfs://ipns/12D3KooWNwPGi9yQs5PyqsL569VUQtQghFNgNug11qZ4WzTAx5VD
 * [new branch]      master -> master
Branch 'master' configured to track the remote 'master' branch of 'origin'.
=> **project created!**

**Thanks a lot for your great initiative!**

**Comment by MeBrei** [comment]
We hope that the two fixes #617 and #620 solved this problem.
Are any of you still experiencing the same issue after updating Radicle?
